### PR TITLE
I have made a wrapper to use GPflow as an external library for Gaussi…

### PR DESCRIPTION
**Pull request by OwenThomas**
…an Processes modelling. It required the modification of the local install of GPflow which is not ideal, so I have given the GPflow a bit of a nudge to perform the necessary update. I also included a hardcoded change in parameter_inference.py, which is definitely a bit crude and can be fixed in the future. I have also edited the BOLFI.py tutorial to BOLFIgpflow.py to provide a minimal example of running a BOLFI instance with the GPflow package.

#### Summary:
Please provide a short summary here

#### Please make sure
- You have updated the CHANGELOG.rst
- You have updated the documentation (if applicable)

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD3 (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: BOLFIgpflow.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
+
+# coding: utf-8
+
+# This tutorial is generated from a [Jupyter](http://jupyter.org/) notebook that can be found [here](https://github.com/elfi-dev/notebooks).
```
Comment (vuolleko) [inline BOLFIgpflow.py]: Great! But please put this notebook in the notebooks repo (or the zoo).

File: elfi/methods/parameter_inference.py [hunk 2]
```diff
@@ -834,8 +835,7 @@ def __init__(self,
         super(BayesianOptimization, self).__init__(
             model, output_names, batch_size=batch_size, **kwargs)
 
-        target_model = target_model or \
-            GPyRegression(self.model.parameter_names, bounds=bounds)
+        target_model = GPflowRegression(self.model.parameter_names, bounds=bounds)
```
Comment (vuolleko) [inline elfi/methods/parameter_inference.py]: The `target_model` must be user-givable (as it was), though the default case can be discussed. :)

File: elfi/methods/bo/gpflow_regression.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
+"""This module contains an interface for using the GPflow library in ELFI."""
+
```
Comment (vuolleko) [inline elfi/methods/bo/gpflow_regression.py]: Hmm... perhaps the current `gpy_regression.py` could be refactored to `gp_regression.py` that includes also this?

File: elfi/methods/parameter_inference.py (hunk 2 above)
Comment (OwenThomas) [reply elfi/methods/parameter_inference.py]: Yes, this was a quick fix on my part: I'm not trying to dethrone GPy from the default.

**Comment by OwenThomas** [comment]
This is not good software practice, but for the minute inserting the following code into /gpflow/model.py should make everything work:

    @AutoFlow((float_type, [None, None]))
    def predict_f_gradients(self, Xnew):
        """
        Compute the gradients of the mean and variance of the latent function(s)
        at the points `Xnew` with respect to the predictive covariates.
        """
        pred_mean, pred_var=self.build_predict(Xnew)
        pred_grad_mean=tf.gradients(pred_mean,Xnew)
        pred_grad_var=tf.gradients(pred_var,Xnew)
        return pred_grad_mean[0], pred_grad_var[0]

**Comment by OwenThomas** [comment]
The relevant GPflow PR can be found here: https://github.com/GPflow/GPflow/pull/247

**Comment by OwenThomas** [comment]
I can do some benchmarking if needed: I have no prior intuition as to which situations GPy or GPflow might be expected to be faster: it is more the model specification functionality in each that I am more interested in.

**Comment by vuolleko** [comment]
No real need for benchmarking; I was just interested to know if this makes e.g. sampling faster.

**Pull request closed without merging**